### ISSUE TITLE
about sim2sim.

### ISSUE THREAD

**Issue by ycwang-123** [SII-FUSC/AME_Locomotion#4]
AME is an impressive piece of work, and I am currently working on sim2sim transfer based on this repository. However, after loading the ame2.pt model in MuJoCo, the G1 robot's joints keep oscillating—it can barely stand still on flat ground and sways continuously. The robot can walk, but only in a stumbling manner. Even without any linear velocity command, the robot wobbles in place.  Thanks for your answering!

**Comment by SII-FUSC** [comment]
Have you conducted the second-stage training with domain randomization enabled?”

**Comment by ycwang-123** [comment]
I haven't completed the training yet. May I ask how many iterations are generally needed for the first and second stages? Currently, I can only train with up to 1024 envs using a 4090. Additionally, for sim2sim, I used the ame2.pt from the /pretrained directory in your codebase. After tuning multiple KPKD parameters, the robot can now go up and down stairs. I plan to retrain with new KPKD values.

**Comment by SII-FUSC** [comment]
You may try the following PD gains and check whether the oscillation still occurs:

kps = [100., 100., 100., 150., 40., 40.,  # Left leg
       100., 100., 100., 150., 40., 40.,  # Right leg
       200., 40., 40.,                    # Waist
       40., 40., 40., 40., 40., 40., 40., # Left arm
       40., 40., 40., 40., 40., 40., 40.] # Right arm

kds = [2., 2., 2., 4., 2., 2.,    # Left leg
       2., 2., 2., 4., 2., 2.,    # Right leg
       5., 5., 5.,                # Waist
       10., 10., 10., 10., 10., 10., 10.,  # Left arm
       10., 10., 10., 10., 10., 10., 10.]  # Right arm

In addition, in our experiments, we used 4096 parallel environments and trained for 10k iterations in each of the two stages.

**Comment by ycwang-123** [comment]
Thank you very much for your reply. I am currently using the PD gains you provided, but in sim2sim the robot's joints oscillate violently and it can barely stand stably. However, there is no such issue in Isaac Lab.

**Comment by SII-FUSC** [comment]
That's strange. In my MuJoCo setup, the robot works normally with this set of PD gains. Have you checked whether the action scale is consistent with the training configuration?

**Comment by ycwang-123** [comment]
> That's strange. In my MuJoCo setup, the robot works normally with this set of PD gains. Have you checked whether the action scale is consistent with the training configuration?

My action_scale is set to 0.25, and the other scales are also aligned with IsaacLab. I found that the issue was caused by a bug in my sim2sim code, which led to the severe jittering. This problem has now been resolved. Thank you very much for your help!

**Comment by KZyred** [comment]
The PD parameters has made me quite curious. Greatly appreciate if you could provide any references or explanations regarding their tuning. I have come across them in various sources, but I am still unsure how they are chosen and what principles guide their selection.

Thank you very much.

**Comment by ycwang-123** [comment]
> The PD parameters has made me quite curious. Greatly appreciate if you could provide any references or explanations regarding their tuning. I have come across them in various sources, but I am still unsure how they are chosen and what principles guide their selection.
> 
> Thank you very much.

Have you read the BeyondMimic paper? It includes a method for calculating KPKD based on motor rotor [https://beyondmimic.github.io/](https://github.com/SII-FUSC/AME_Locomotion/issues/url)

**Comment by KZyred** [comment]
I have taken a quick look at the papers and found some interesting information. I'll spend more time exploring them and implementing the relevant ideas.
Thank you very much.

- `Actions and Joint-Level Control`, `Nominal Armature Model` - `ZEST: Zero-shot Embodied Skill Transfer for Athletic Robot Control`
- `Heuristically Designed Paramters` - `BeyondMimic: From Motion Tracking to Versatile Humanoid Control via Guided Diffusion`
- `Gain Tuner Extension` -`https://docs.isaacsim.omniverse.nvidia.com/6.0.0/robot_setup/ext_isaacsim_robot_setup_gain_tuner.html#gain-tuner-extension`

**Comment by KZyred** [comment]
<img width="1034" height="467" alt="Image" src="https://github.com/user-attachments/assets/6743f800-6d19-4770-94c3-4b9c84a2b6f7" />

@SII-FUSC @ycwang-123 
I have another question regarding the reward weights. Why are there seemingly "magic numbers" like these? I understand that changing one weight can affect the balance among the different reward terms, but I'm curious about how these values were originally chosen.

For example, why is the weight for action-rate set to 5.0e-3, while the weight for joint position limits is much larger (1.0)? What is the rationale behind such a significant difference in scale?
Why isn't the action-rate weight set to 1.0 and the joint position limits weight set to 5.0 instead?

**Comment by ycwang-123** [comment]
> <img alt="Image" width="1034" height="467" src="https://private-user-images.githubusercontent.com/86369121/601949528-6743f800-6d19-4770-94c3-4b9c84a2b6f7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODA1NTE4ODYsIm5iZiI6MTc4MDU1MTU4NiwicGF0aCI6Ii84NjM2OTEyMS82MDE5NDk1MjgtNjc0M2Y4MDAtNmQxOS00NzcwLTk0YzMtNGI5Yzg0YTJiNmY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjA0VDA1Mzk0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiOTY3YjAxNjNlMzFmODllMTc4Zjg4MDNkODhjNWMwNjQxZWMxMTc3MDA5M2Y3OTQwMzE2YTgxN2U0ZTVmMjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.px1XrCkATUJodlr9P0vKnUID3P7y1BOjOsIF-dQgXmg">
> [@SII-FUSC](https://github.com/SII-FUSC) [@ycwang-123](https://github.com/ycwang-123) I have another question regarding the reward weights. Why are there seemingly "magic numbers" like these? I understand that changing one weight can affect the balance among the different reward terms, but I'm curious about how these values were originally chosen.
> 
> For example, why is the weight for action-rate set to 5.0e-3, while the weight for joint position limits is much larger (1.0)? What is the rationale behind such a significant difference in scale? Why isn't the action-rate weight set to 1.0 and the joint position limits weight set to 5.0 instead?

I guess it's because someone tried it before and got undesirable training results. These coefficients are all empirically derived.

**Comment by jie0110** [comment]
@ycwang-123  Hello, could you please share some sim2sim code examples with me? Thank you so much!
